cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go

```diff
@@ -1132,3 +1133,207 @@ func TestNodeGroupWithFailedMachine(t *testing.T) {
+					"nodeGroupLabel":          "value",
+					"anotherLabel":            "anotherValue",
+				},
+			},
+		},
+		{
+			name: "When the NodeGroup can scale from zero and the Node still exists, it includes the known node labels",
+			nodeGroupAnnotations: map[string]string{
+				memoryKey: "2048",
+				cpuKey:    "2",
+			},
+			config: testCaseConfig{
+				includeNodes: true,
+				expectedErr:  nil,
+				nodeLabels: map[string]string{
+					"kubernetes.io/os":                 "windows",
+					"kubernetes.io/arch":               "arm64",
+					"node.kubernetes.io/instance-type": "instance1",
+					"anotherLabel":                     "nodeValue", // This should not be copied as it is not a well known label
+				},
+				nodegroupLabels: map[string]string{
+					"nodeGroupLabel": "value",
+					"anotherLabel":   "anotherValue",
+				},
+				expectedCapacity: map[corev1.ResourceName]int64{
+					corev1.ResourceCPU:        2,
+					corev1.ResourceMemory:     2048 * 1024 * 1024,
+					corev1.ResourcePods:       250,
+					gpuapis.ResourceNvidiaGPU: 0,
+				},
+				expectedNodeLabels: map[string]string{
+					"kubernetes.io/os":                 "windows",
+					"beta.kubernetes.io/os":            "linux",
+					"kubernetes.io/arch":               "arm64",
+					"beta.kubernetes.io/arch":          "amd64",
+					"nodeGroupLabel":                   "value",
+					"anotherLabel":                     "anotherValue",
+					"node.kubernetes.io/instance-type": "instance1",
+				},
+			},
+		},
+	}
+
+	test := func(t *testing.T, testConfig *testConfig, config testCaseConfig) {
+		if testConfig.machineDeployment != nil {
+			testConfig.machineDeployment.Spec.Template.Spec.SetLabels(config.nodegroupLabels)
+		} else {
+			testConfig.machineSet.Spec.Template.Spec.SetLabels(config.nodegroupLabels)
+		}
+
+		if config.includeNodes {
+			for i := range testConfig.nodes {
+				testConfig.nodes[i].SetLabels(config.nodeLabels)
+			}
+		} else {
+			testConfig.nodes = []*corev1.Node{}
+		}
+
+		controller, stop := mustCreateTestController(t, testConfig)
+		defer stop()
+
+		nodegroups, err := controller.nodeGroups()
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+
+		if l := len(nodegroups); l != 1 {
+			t.Fatalf("expected 1 nodegroup, got %d", l)
+		}
+
+		ng := nodegroups[0]
+		nodeInfo, err := ng.TemplateNodeInfo()
+		if config.expectedErr != nil {
+			if err != config.expectedErr {
+				t.Fatalf("expected error: %v, but got: %v", config.expectedErr, err)
+			}
+			return
+		}
+
+		for resource, quantity := range nodeInfo.Node().Status.Allocatable {
+			expectedCapacity, ok := config.expectedCapacity[resource]
+			if !ok {
+				continue
+			}
+			if quantity.Value() != expectedCapacity {
+				t.Errorf("Expected allocatable %q: %+v, Got: %+v", resource, expectedCapacity, quantity.Value())
+			}
+		}
+
+		for resource, quantity := range nodeInfo.Node().Status.Capacity {
+			expectedCapacity, ok := config.expectedCapacity[resource]
```

Comment:
It seems that we shoud test that all `expectedCapacity` labels are included on the Node resource as well.